core/src/main/java/io/aiven/kafka/tieredstorage/chunkmanager/cache/DiskBasedChunkCache.java

```diff
@@ -0,0 +1,119 @@
+
+import static java.nio.file.StandardCopyOption.ATOMIC_MOVE;
+
+public class DiskBasedChunkCache extends ChunkCache<Path> {
+
+    private static final Logger log = LoggerFactory.getLogger(DiskBasedChunkCache.class);
+    static final String TEMP_CACHE_DIRECTORY = "temp";
+    static final String CACHE_DIRECTORY = "cache";
+
+    private DiskBasedChunkCacheConfig config;
+
+    public DiskBasedChunkCache(final ChunkManager chunkManager) {
+        super(chunkManager);
+    }
+
+    @Override
+    public InputStream cachedChunkToInputStream(final Path cachedChunk) {
+        try {
+            return Files.newInputStream(cachedChunk);
+        } catch (final IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Writes the chunk to specified directory on disk.
+     * To be able to understand if the files are fully written to disk during a potential recovery process,
+     * this implementation initially writes them to temporary directory and then atomically copies them
+     * to the actual caching directory.
+     */
+    @Override
+    public Path cacheChunk(final ChunkKey chunkKey, final InputStream chunk) throws IOException {
+        final Path tempChunkPath = config.baseCachePath().resolve(TEMP_CACHE_DIRECTORY)
+                .resolve(chunkKey.uuid.toString())
+                .resolve(String.valueOf(chunkKey.chunkId));
+        if (Files.notExists(tempChunkPath.getParent())) {
+            Files.createDirectories(tempChunkPath.getParent());
+        }
+        final Path tempCached = writeToDisk(chunk, tempChunkPath);
+        log.info("Chunk file has been stored to temporary caching directory {}", tempCached);
+        final Path cachedChunkPath = config.baseCachePath()
+                .resolve(CACHE_DIRECTORY).resolve(chunkKey.uuid.toString())
+                .resolve(String.valueOf(chunkKey.chunkId));
+        if (Files.notExists(cachedChunkPath.getParent())) {
+            Files.createDirectories(cachedChunkPath.getParent());
+        }
+        final Path newPath = Files.move(tempCached, cachedChunkPath, ATOMIC_MOVE);
+        log.debug("Chunk file has been moved to cache directory {}", newPath);
+        return newPath;
+    }
+
+    private static Path writeToDisk(final InputStream chunk, final Path tempChunkPath) throws IOException {
+        try (chunk) {
+            return Files.write(tempChunkPath, chunk.readAllBytes());
+        }
+    }
+
+    @Override
+    public RemovalListener<ChunkKey, Path> removalListener() {
+        return (key, path, cause) -> {
+            try {
+                Files.delete(path);
+            } catch (final IOException e) {
+                log.error("Failed to delete cached file for key {} with path {} from cache directory."
+                              + " The reason of the deletion is {}", key, path, cause, e);
+            }
+            log.info("Deleted cached file for key {} with path {} from cache directory."
```

Comment:
In case of error, the logging is incorrect. I think the success logging should be inside the `try`.